### DISCUSSION TITLE
some ideas about motion state

### DISCUSSION THREAD

**Discussion by jixiuf** [meow-edit/meow#681] [Ideas]
Regarding motion state, I have some thoughts.

For example, in help-mode, by default it belongs to motion state. However, I want to use some default normal state keybindings in help-mode instead of those in help-mode-map. For instance, I would like to bind "r" in help-mode-map to `#'help-go-forward`.

I considered executing the following code to achieve this:
```elisp
(keymap-unset help-mode-map "r" t)
```
However, the actual effect is that "r" becomes undefined.

Therefore, I want to understand if it is possible to set the priority of the following keymaps in motion state: motion-state-map > help-mode-map > normal-state-map. This way, if I want the keys from normal-state-map to be effective, I only need to unbind them in the major-mode, while the latest priority remains in motion-state-map, which is usually empty.

Thus, my idea is to let the major-mode of a mode in motion state automatically inherit from `meow-normal-state-keymap`. This can be done using the following macro, and meow just needs to call `meow-set-keymap-parent` at the appropriate time:
```elisp
(defmacro meow-set-keymap-parent (map &optional parent)
  `(set-keymap-parent ,map (make-composed-keymap (keymap-parent ,map)
                                                 (or ,parent meow-normal-state-keymap))))
```
After loading help mode, I can execute the following code:
```elisp
(with-eval-after-load 'help
  (meow-set-keymap-parent help-mode-map))
```

Users would only need to do the following to achieve the desired effect:
```
(keymap-unset help-mode-map "r" t)
```
--------

  关于 motion state，我有一些想法。

以 help-mode 为例，默认情况下 help-mode 属于 motion state。然而，我想在 help-mode 下使用某些默认的 normal state 按键绑定，而不是 help-mode-map 中的绑定。例如，我希望将 help-mode-map 中的 "r" 绑定为 #'help-go-forward。

我考虑执行以下代码来实现：
```elisp
(keymap-unset help-mode-map "r" t)
```
然而，实际效果是 "r" 变成了 undefined。

由此，我想要了解是否可以在 motion state 下，设置以下几个 keymap 的优先级：motion-state-map > help-mode-map > normal-state-map。这样，如果想让 normal-state-map 的键生效，只需在 major-mode 中取消绑定即可，而最新的优先级放在 motion-state-map 内，通常 motion-state-map 是空的。

因此，我的想法是让处于 motion state 的 mode 其 major-mode-map 自动继承自 meow-normal-state-keymap。可以通过以下宏来实现，meow 只需在合适的时机调用 meow-set-keymap-parent：
```elisp
(defmacro meow-set-keymap-parent (map &optional parent)
  `(set-keymap-parent ,map (make-composed-keymap (keymap-parent ,map)
                                                 (or ,parent meow-normal-state-keymap))))
```
在 help mode 加载后执行以下代码：
```elisp
(with-eval-after-load 'help
  (meow-set-keymap-parent help-mode-map))
```

用户只需要如下做就能实现想要的效果。

```
(keymap-unset help-mode-map "r" t)
```

**Comment by jixiuf** [comment]
```
(defmacro meow-set-keymap-parent (map-or-mode &optional parent)
  "Set the parent keymap for MAP-OR-MODE to PARENT.

MAP-OR-MODE can be a map or a mode
Returns PARENT. PARENT should be nil or another keymap.
Default is 'meow-normal-state-keymap' when PARENT is nil."
  `(let ((map (if (keymapp ,map-or-mode)
                  ,map-or-mode
                (let ((map-name (intern (concat (symbol-name ,map-or-mode) "-map"))))
                  (when (boundp map-name)
                    (symbol-value map-name))))))
     (when (keymapp map)
       (set-keymap-parent map (make-composed-keymap (keymap-parent map)
                                                    (or ,parent meow-normal-state-keymap))))))

(defvar meow-motion-parent-keymaps (make-hash-table :test #'equal))
(defun meow-motion-set-keymap-parent()
  (when (and meow-motion-mode
             (not (gethash major-mode meow-motion-parent-keymaps)))
    (puthash major-mode t meow-motion-parent-keymaps)
    (meow-set-keymap-parent major-mode meow-normal-state-keymap)))

(add-hook 'meow-motion-mode-hook #'meow-motion-set-keymap-parent)

```
After further improvement, when entering motion state, the major-mode map will automatically have its parent set to the normal state. Another advantage is, for example in help-mode, the characters "," and "." are not bound by default. By enabling this feature, the functions for "," and "." remain available in help-mode under moeow.

进一步做了改进后，当进入 motion state 后 自动将 major-mode map 的parent 设置为 normal state
还有一个好处是 以help-mode 为例， help-mode 下并没有对",","." 进行绑定,
启用此功能后， help-mode 下 meow 的 ",","." 功能依然能用

你觉得这个功能怎么样，有没有可能合到meow中。

**Comment by DogLooksGood** [comment]
After some thinking, I don't think we need this in meow.  

First of all, this could break some configurations which already uses the parent map.  

Second, there's no obvious advantage to do this setup in meow than do it without meow.  I can always use this tricky to achieve, and it's already straightforward. If I want `help-mode-map` to inherit the `meow-normal-map`, I just set it. 

I could have a keymap which supports HJKL movements, and use it as the parent map at anywhere I want.

**Reply by jixiuf** [reply]
> After some thinking, I don't think we need this in meow.

it's ok for me.
> 
> First of all, this could break some configurations which already uses the parent map.

It won't , it use `make-composed-keymap`  in `meow-set-keymap-parent` ,
The `meow-normal-map` will be an ancestor of the `help-mode-map`, not just a parent, which means that the =meow-normal-map= will have the lowest priority.

`meow-normal-map` 将会是`help-mode-map` 的祖先，而不仅仅是parent,所以`meow-normal-map` 它的优先级最低，
> 
> Second, there's no obvious advantage to do this setup in meow than do it without meow. I can always use this tricky to achieve, 
and it's already straightforward. If I want `help-mode-map` to inherit the `meow-normal-map`, I just set it.

The advantages is that it allows all modes of the motion state to uniformly inherit from meow-normal-map. This means that motion state can automatically adopt the keymap settings from the normal state, eliminating the need for users to set them individually.

它的优点正是为motion state的所有mode 让其统一继承自meow-normal-map.从而让motion state 能继承normal state中的keymap设置,用户不必费心一一设置
> 
> I could have a keymap which supports HJKL movements, and use it as the parent map at anywhere I want.

**Reply by DogLooksGood** [reply]
I personally don't think inheriting from `meow-normal-state-keymap` is a good idea, as this keymap is obviously for text editing. That's why we have motion state for where we don't need text editing. Using the text editing keymap as a fallback for a non-text editing mode looks weird to me. If users want a fallback for where a single key navigation is needed, they should build another simple keymap and use it.
